[WIP] bu12.1 Branch - Parallel Block Validation

ptschip opened the pull request:
This is the updated branch which is stable as far as I know.  It still needs the cs_main lock however I'm working on another branch which will remove that dependency.

The very first commit is everything that was in the original branch.  Many changes, most notable are:

1)  Moving code into parallel.cpp/parallel.h
2)  Removing the need for the extra scoped lock
3)  Using only one semaphore instead of two.  Using the same semaphore for IBD as well as PV.
4)  Consolidating the scriptcheckqueue thread pool initialization  and putting everything into parallel.cpp and simplifying the config such that we can easily add or subtract thread pools with a simple change.

Currently the parallel.py script does not work if you are in debug mode.  But otherwise, It generally runs fine but still occasionally fails because the small block showed up too late and didn't win the race.  I have to find a better way to create a much longer to validate block such that the timing of block download is not an issue.

Commit:
Parallel Block Validation

- add 3 more script check queues
- use a wrapper function to call HandleBlockMessageThread
- Create semaphores to control and handle queue selection and throughput
- Update locking - we do not need to hold cs_main when checking inputs
- Flush the temporary view to the base view rather than re-running every txn with UpdateCoins
- unlock cs_main before waiting for script check threads to complete
- Python test for Parallel Validation

- Ability to Quit scriptcheck threads

When  4 blocks are running in parallel and a
new, smaller block shows up, we need to be able to interrupt the
script threads that are currently validating for a one of the  blocks
so that we can free up a script check queue for the new smaller  block.

- Update the nBlockSequenceId after the block has advanced the tip

This is important for Parallel Validation.  By decreasing the sequence id
we are indicating that this block represents the current pindexMostWork. This
prevents the losing block from having the pindexMostWork point to it rather
than to the winning block.

- Continously check for new blocks to connect

With Parallel Validation new blocks can be accepted while we are connecting at
tip therefore we need to check after we connect each block whether a potentially
longer more work chain now exists. If we don't do this then we can at times end
up with a temporarily unconnected block until the next block gets mined and
another attempt is made to connect the most work chain.

- Terminate any PV threads during a re-org

PV can only operate with blocks that will advance the current chaintip (fork1). Therefore,
if we are needing to re-org to another chain (fork 2)  then we have to kill any currently
running PV threads assoicated with the current chain tip on fork1. This is the
solution to the problem of having two forks being mined while there is an attack
block processing on fork1. If fork2 continues to be mined and eventually pulls
ahead of fork1 in proof of work, then a re-org to fork2 will be initiated causing
the PV threads on fork1 to be terminated and fork2 blocks then connected and fork2 then
becoming the chain active tip.

- Use Chain Work instead of nHeight to self terminate a PV thread

If the Chain Work has changed either positve or negative to where it
was when we started the PV thread then we will exit the thread.  Previously
we were using Chain Height, which worked fine, but this is more understantable
from a coding perspective and also we added the feature to check for when
Chain Work has decreased from the starting point which would indicate that
a re-org was underway.

-Move ZMQ notifications to ActivateBestChainStep

We must notify ZMQ after each tip is connected rather
than after ActivateBestChain otherwise we can miss a block
when there are several to connect together at once.

This particularly seems to help sync issues on regest where
we're mining many blocks all at the same time.

-Completely remove cs_main locks from checkinputs and sigs during PV

During the loop whre we check inputs and signatures we can completely
remove the cs_main locks by making sure to add an internal lock
cs_main lock to ChainWorkHasChanged().  This function rarely will
get invoked and only if the thread is about to exit so there is no
problem here of adding any additional overhead.

Removing the cs_main locks here is the most important step in
completely removing any significant inter-dependancy between
parallel validation threads.

-Create SetLocks() used to set the locking order when returning from PV

If there is an error during PV then we have to make sure the locks
are set in the right order before returning.  cs_main must always
be locked when we return from ConnectBlock() as cs_main is recursive
but we must also ensure that the scriptlock is unlocked before
re-locking cs_main to avoid a potential deadlock.

By creating SetLocks() we can abstract away the setting of the locking
order and prevent any developer confusion or possible introduction of  errors
into the code if future changes are made in the ConnectBlock() function.

ptschip committed:
Consolidate and simplify the sript check thread pool creation

In the past the code for creaating the script check threads and pools
was all over the place in several files but is now consolidated in
parallel.cpp and parallel.h.  Also is is much easier to make any
changes to the number of scriptcheckqueue's by just editing two
lines in  parallel.cpp.

ptschip committed:
Block handling updates

Use only one semaphore for PV instead of two.

Move cs_blocksemaphore into globals.

Assign a scriptqueue to a running thread

Do not process a duplicate block

check whether chainwork has advanced when we initialize PV

refactor handblockmessage and put in parallel.cpp

QT changes

-Add StartShutdown when closing QT Gui

StartShutdown() was never getting called which sets the shutdown
flag to true.  Generally this would be rare to cause a problem  but in PV
it can become an issue fairly easily if a user wants
to shutdown while many blocks are being connected during IBD.

-Correctly notify the UI of block tip changes

When many blocks are being connected during IBD it was
possible the the UI was not getting updated.  This is a Core
bug which happened rarely but in PV it can happen frequently.

update IsChainNearlySyncdInit after each block is processed

Rebase fixes

Run with PV when doing IBD as well as new blocks

Move cs_blockvalidationthread to parallel.h

Only print out error message when state.Invalid() or state.IsError()

When activatebestchain fails during PV due to a competing block being
beaten, we don't want to print out an error because it really isn't an
error. The block validation was terminated but the block will be store
on disk for the future in the event of a re-org.

PV.SetLocks() for SequenceLocks

remove cs_updateutxo

Do not allow two thinblocks to be validating at the same time

interconnect_nodes in parallel.py

ptschip committed:
Add Read/Write locks to UTXO cache

During PV we may have to access the UTXO concurrent with other
threads and need to ensure proper locking.  This is just an
additional layer of isolation since LevelDB has it's own locking.
As such, these locks will probably be temporary until we can be
sure that PV is stable.

Commit:
More block handling updates

-Change where we quit all competing threads

We must quit the competing threads before updating the UTXO
otherwise we may get a DOS from another thread when it finds
that it's inputs have been spent but before the chaintip has
had a chance to update.

- updates for parallely.py for python3

- fix p2p-versionbits-warning for python3

- Update the UI if we are close to being syncd

Update the block Current Number of Blocks in the UI if we
are close to finishing our sync. This way when users turn off
their node for a while and turn it on they can see each block
being added as it finishes processing.

-Track whether the current block is acutally validating inputs

We need to do this in the event that two of the same block arrive
and have launched validation threads.  If this should happen then
one of those blocks will begin checking inputs before the other and
we can set a flag which indicates that fact.  Then when the other
block beginns checking inputs it can check to see if the same
block is also currently validatin and if so then exit.

-Use nChainWork to determine whether we have any work to do

When a newblock arrives we check whether the chaintip already matches
pindexMostWork.  However in PV, pindexMostWork may not necessarily point
to the chaintip, therefore we use pindexMostWork->nChainWork to determine
if an attempt should be made to connect this block.

ptschip commented:
This version still requires the cs_main locks when checking inputs however I'm getting close to finishing another branch which should replace this one and does not require the locks.

Also, I'd like to cleanup connectblock() further and turn it into an object so that we can use the destructor to set the recursive locks correctly on return or some sort of exception.  This should be much safer and more maintainable.

ptschip commented:
I pushed up the new branch where we take out the cs_main locks and make the UTXO thread safe.

One thing I noticed in passing was that AsserLockHeld() is not threadsafe and neither does it track which thread has a lock held, so one thread could have a lock and another thread will think it also has the lock.  I'll work on that and submit a separate PR for it.

ptschip committed:
Simplify ConnectBlock and remove unnecessary if(fParallel) statments

Some updates to parallel.py

mine blocks with parallel off

deadalnix commented:
I suggest we start merging the part of this PR that can be isloated on their own in the dev branch. This'll allow travis to run on them and will make sure nobody is developing something that is incompatible with these.

I'm going over this and will point part that can be isolated in their own PR and merged today.

deadalnix commented on qa/rpc-tests/excessive.py:
```diff
@@ -525,7 +525,6 @@ def testExcessiveBlockSize(self):
             except JSONRPCException: # could be spent all the txouts
               pass
           logging.debug("%d transactions" % ntxs)
-          time.sleep(1)
```
What's the reason for this ? If this is trully useful, then it can go in its own PR, if not, then maybe it just needs to be discarded.

ptschip replied:
i don't see that in the file...it must have been take out in a later commit.

deadalnix commented on qa/rpc-tests/maxblocksinflight.py:
```diff
@@ -13,8 +13,10 @@
+BU: We allow for 256 blocks in flight with 32 from a single peer
 '''
-MAX_REQUESTS = 128
+MAX_REQUESTS = 256
```
This can probably go in in its own PR.

ptschip replied:
this was put in a later PR which is merged now in the "release" branch

deadalnix commented on qa/rpc-tests/p2p-acceptblock.py:
```diff
@@ -278,8 +278,15 @@ def run_test(self):
+        while self.nodes[0].getblockcount() != 290:
+            time.sleep(1)
+            j = j + 1
+            if (j > 60):
```
sleeping is not an accurate time measure. A better way to do this is to get a timestamp from the system and then compare with the current timestamp. If the difference is greater than 60, then bail out.

Otherwise, the change to this test can go in in their own PR today.

ptschip replied:
no it's not 100% accurate but for the purpose of what we're trying to do it's fine IMO.  It doesn't matter if it's exactly 60 seconds , it's just there so the test can at some point finish and proceed without hanging up.

deadalnix commented on qa/rpc-tests/p2p-versionbits-warning.py:
```diff
@@ -90,6 +90,7 @@ def send_blocks_with_version(self, peer, numblocks, nVersionToUse):
             block_time += 1
             height += 1
             tip = block.sha256
+            #tip = block.hash
```
Why this change ?

ptschip replied:
oops , that's wasn't a change, it's just one of my comments i had left in there when i was working on a bug.

ptschip replied:
fixed now.

deadalnix commented on qa/rpc-tests/p2p-versionbits-warning.py:
```diff
@@ -112,27 +113,39 @@ def run_test(self):
         # blocks signaling some unknown bit.
         nVersion = VB_TOP_BITS | (1<<VB_UNKNOWN_BIT)
-        self.send_blocks_with_version(test_node, VB_THRESHOLD-1, nVersion)
+        #self.send_blocks_with_version(test_node, VB_THRESHOLD-1, nVersion)
```
Please do not comment out code you want to remove. The old code remain in source control history so this is not necessary.

ptschip replied:
actually that's true now but historically we always left commented out code in place because otherwise we could have issues during rebases on Core...However, now that we're not rebasing anymore these comments can come out.   Thanks for noticing.

fixed.

deadalnix commented on qa/rpc-tests/p2p-versionbits-warning.py:
```diff
@@ -112,27 +113,39 @@ def run_test(self):
         # 3. Now build one period of blocks with >= VB_THRESHOLD blocks signaling
         # some unknown bit
-        self.send_blocks_with_version(test_node, VB_THRESHOLD, nVersion)
+        #self.send_blocks_with_version(test_node, VB_THRESHOLD, nVersion)
```
dito

ptschip replied:
fixed.

deadalnix commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -184,7 +184,8 @@ def initialize_datadir(dirname, n,bitcoinConfDict=None,wallet=None):
     defaults = {"server":1, "discover":0, "regtest":1,"rpcuser":"rt","rpcpassword":"rt",
-                "port":p2p_port(n),"rpcport":str(rpc_port(n)),"listenonion":0,"maxlimitertxfee":0}
+                "port":p2p_port(n),"rpcport":str(rpc_port(n)),"listenonion":0,"maxlimitertxfee":0, "debug":"net",
+                "debug":"parallel", "debug":"thin", "debug":"blk", "debug":"req", "debug":"mempool"}
```
One KV pair per line would be much more readable. We'd also get an history that is much easier to understand as it'll be easy to see who changed what parameter, and in what context instead of seeing a bulk change of all parameters.

ptschip replied:
you're right it would be more readable but maybe should go in a separate PR.

deadalnix commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -603,6 +604,10 @@ def split_transaction(node, prevouts, toAddrs, txfeePer=DEFAULT_TX_FEE_PER_BYTE,
     finally:
       decimal.getcontext().prec = decContext
 
+def assert_not_equal(thing1, thing2):
```
This doesn't looks like it is specific to PV, and should probably go in its own PR.

ptschip replied:
I need this one for the PV regression test to work...it's not used by any other script.

deadalnix commented on src/coins.cpp:
```diff
@@ -162,6 +162,7 @@ void CCoinsViewCache::SetBestBlock(const uint256 &hashBlockIn) {
 
 bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn) {
     assert(!hasModifier);
+
```
While this is a formatting change I'd agree with, please keep formatting change in their own PR. Consider this: if for some reason PV needs to be reverted, we may end up with a conflict here for no reason, which would make the revert much easier.

The reason I'm bringing easy revert here is that you have 2 ways to handle risk: taking measure to make failing rarer, or taking measures to make failing cheaper. In most engineering fields, the former is usually preferable: crashing a building or an airplane is VERY expensive. In software, with the right set of practices, 2 can be very much preferable as it allows to move faster.

Making things easier to revert is one of these practices. in practice, it translate by very focused diffs rather than huge set of changes in their own branches to be merged all at once. Each one of these small change can be reverted individually if there is a problem with it, with little risk of conflict.

deadalnix commented on src/coins.h:
```diff
@@ -74,6 +74,7 @@
 class CCoins
 {
 public:
+
```
dito

ptschip replied:
fixed.

deadalnix commented on src/coins.h:
```diff
@@ -316,6 +317,8 @@ struct CCoinsStats
 {
 public:
+
+
```
dito

ptschip replied:
fixed.

deadalnix commented on src/coins.h:
```diff
@@ -381,6 +384,7 @@ class CCoinsModifier
 class CCoinsViewCache : public CCoinsViewBacked
 {
 protected:
+
```
dito

ptschip replied:
fixed.

deadalnix commented on src/coins.h:
```diff
@@ -396,6 +400,7 @@ class CCoinsViewCache : public CCoinsViewBacked
     mutable size_t cachedCoinsUsage;
 
 public:
+
```
dito

ptschip replied:
fixed.

deadalnix commented on src/main.cpp:
```diff
@@ -2335,12 +2338,16 @@ void static FlushBlockFile(bool fFinalize = false)
+//void ThreadScriptCheck() {
+//    RenameThread("bitcoin-scriptch");
+//    scriptcheckqueue.Thread();
+//}
```
???

ptschip replied:
fixed...we don't need out old comments anymore since we're not rebasing on Core anymore.

deadalnix commented on src/main.cpp:
```diff
@@ -2457,8 +2464,12 @@ static int64_t nTimeIndex = 0;
 static int64_t nTimeTotal = 0;
 
-bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pindex, CCoinsViewCache& view, bool fJustCheck)
+bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pindex, CCoinsViewCache& view, bool fJustCheck, bool fParallel)
```
Taking bool as argument is generally recognized as an anti-pattern. See http://blog.iannelson.systems/back-to-basics-on-the-use-and-abuse-of-the-humble-boolean/ for instance.

I think this part of the code needs some refactoring before PV is introduced. For instance, extract the part of the block validation that are common between PV and regular validation in their own function, and have a base method which is different and handles the locking differently.

ptschip replied:
I agree, I don't' like it that much either but IMO it will require a much larger refactor beyond the scope of this PR.  The underlying problem is with Processnewblock, calling AcitvateBestChain,, calling Activatebestchainstep, calling ConnectTip, calling ConnectBlock()...it really all should be in a class (in fact all of main.cpp needs a refactor similar to what Matt did recently for Core) which would make life much easier but I think that should wait until later otherwise this PR  is just going to get bigger than it already is.  I have it in mind to change all this but as I said I think it should wait, and there is nothing that will change regarding the functionality so I think we should leave this for now.

deadalnix commented on src/main.cpp:
```diff
@@ -2515,20 +2526,20 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
+            }
 	}
-      }
+    }
```
Same comment as before when it comes to formatting.

deadalnix commented on src/main.cpp:
```diff
@@ -2607,48 +2661,84 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
-            control.Add(vChecks);
         }
-
+      
```
One more formatting change. Move in its own PR and/or remove.

ptschip replied:
done

deadalnix commented on src/main.cpp:
```diff
@@ -3128,31 +3252,61 @@ static void PruneBlockIndexCandidates() {
  * pblock is either NULL or a pointer to a CBlock corresponding to pindexMostWork.
  */
-static bool ActivateBestChainStep(CValidationState& state, const CChainParams& chainparams, CBlockIndex* pindexMostWork, const CBlock* pblock)
+static bool ActivateBestChainStep(CValidationState& state, const CChainParams& chainparams, CBlockIndex* pindexMostWork, const CBlock* pblock, bool fParallel)
```
bool param, require refactoring first.

deadalnix commented on src/main.cpp:
```diff
@@ -3049,13 +3172,14 @@ static CBlockIndex* FindMostWorkChain() {
             if (depth < excessiveAcceptDepth)
 	      {
 		fRecentExcessive |= ((pindexTest->nStatus & BLOCK_EXCESSIVE) != 0);  // Unlimited: deny this candidate chain if there's a recent excessive block
+if ((pindexTest->nStatus & BLOCK_EXCESSIVE) != 0) LogPrintf("recent excessive block\n");
```
Formating.

ptschip replied:
oops that shouldn't be there. fixed.

deadalnix commented on src/main.cpp:
```diff
@@ -3049,13 +3172,14 @@ static CBlockIndex* FindMostWorkChain() {
 	      }
 
-            if (fFailedChain | fMissingData | fRecentExcessive) break;
+            if (fFailedChain | fMissingData | fRecentExcessive) { LogPrintf("breaking ######\n");break;}
```
Please format block properly. I assume this is something here for debug and you plan to remove it, feel free to ignore if so.

ptschip replied:
same as before...taken out...don't need it anymore.

deadalnix commented on src/main.cpp:
```diff
@@ -3128,31 +3252,61 @@ static void PruneBlockIndexCandidates() {
         // Don't iterate the entire list of potential improvements toward the best tip, as we likely only need
         // a few blocks along the way.
-        int nTargetHeight = std::min(nHeight + 32, pindexMostWork->nHeight);
+        int nTargetHeight = std::min(nHeight + (int)BLOCK_DOWNLOAD_WINDOW, pindexMostWork->nHeight);
```
Why do you need a cast ?

ptschip replied:
because block_download_window is unsigned and if we don't we get compiler warnings on some systems.

deadalnix commented on src/main.cpp:
```diff
@@ -3128,31 +3252,61 @@ static void PruneBlockIndexCandidates() {
+        int nTargetHeight = std::min(nHeight + (int)BLOCK_DOWNLOAD_WINDOW, pindexMostWork->nHeight);
         vpindexToConnect.clear();
-        vpindexToConnect.reserve(nTargetHeight - nHeight);
+        //vpindexToConnect.reserve(nTargetHeight - nHeight);
```
Do not comment code. Source control is your friend.

deadalnix commented on src/main.cpp:
```diff
@@ -3206,8 +3391,9 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
  * that is already loaded (to avoid loading it again from disk).
  */
-bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams, const CBlock *pblock) {
+bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams, const CBlock *pblock, bool fParallel) {
```
bool param, refactoring.

deadalnix commented on src/main.cpp:
```diff
@@ -3217,15 +3403,57 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
 
+            // This is needed for PV because FindMostWorkChain does not necessarily return the block with the lowest nSequenceId
+            if (fParallel)
+            {
```
egyptian braces and new line braces are both used in this codebase. I think we should strive for consistency. While semantic is the same in C++, semantic differs in JS and we are using it for the website part, so I'd advocate for egyptian braces.

deadalnix commented on src/main.cpp:
```diff
@@ -3823,7 +4052,7 @@ static bool IsSuperMajority(int minVersion, const CBlockIndex* pstart, unsigned
 
 
-bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, const CNode* pfrom, const CBlock* pblock, bool fForceProcessing, CDiskBlockPos* dbp)
+bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, const CNode* pfrom, const CBlock* pblock, bool fForceProcessing, CDiskBlockPos* dbp, bool fParallel)
```
bool, refactoring first, nothing that hasn't been said already.

deadalnix commented on src/main.cpp:
```diff
@@ -4588,11 +4820,11 @@ void static CheckBlockIndex(const Consensus::Params& consensusParams)
+           //         assert(setBlockIndexCandidates.count(pindex));
+                    // If some parent is missing, then it could be that this block was in
+                    // setBlockIndexCandidates but had to be removed because of the missing data.
+                    // In this case it must be in mapBlocksUnlinked -- see test below.
```
Please no commented out code.

deadalnix commented on src/main.cpp:
```diff
@@ -5007,6 +5239,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         CAddress addrMe;
         CAddress addrFrom;
         uint64_t nNonce = 1;
+
```
formating change.

ptschip replied:
fixed.

deadalnix commented on src/main.cpp:
```diff
@@ -5144,8 +5377,9 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-                pfrom->PushMessage(NetMsgType::SENDHEADERS);
+            // BU: Parallel Block Validation.  We must have all headers first announcements turned off.
+            //if (!IsThinBlocksEnabled())
+            //    pfrom->PushMessage(NetMsgType::SENDHEADERS);
```
Commented out code.

ptschip replied:
this will be re-enabled at some point so need to keep it in.

deadalnix commented on src/main.cpp:
```diff
@@ -5795,9 +6025,10 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         int missingCount = 0;
         int unnecessaryCount = 0;
+        {
+        LOCK2(cs_main, cs_xval);
```
Ident.

ptschip replied:
this is a matter of style..we see this throughout Bitcoin..,when locking we don't always indent.  I don't really mind either way but sometimes it's easier to read the code or is less code change.

deadalnix commented on src/main.cpp:
```diff
@@ -6688,7 +6914,7 @@ bool SendMessages(CNode* pto)
 
                 // BU - here we only want to forward message inventory if our peer has actually been requesting useful data or
-                //      giving us useful data.  We give them 10 minutes to be useful but then choke off their inventory.  This
+                //      giving us useful data.  We give them 2 minutes to be useful but then choke off their inventory.  This
```
This looks like a change that can go in its own PR.

deadalnix commented on src/main.h:
```diff
@@ -429,37 +432,35 @@ class ValidationResourceTracker
+//    }
+
+//    ScriptError GetScriptError() const { return error; }
+//};
```
commented out code.

ptschip replied:
done

deadalnix commented on src/net.cpp:
```diff
@@ -904,14 +907,12 @@ class CNodeRef {
 };
 
-#if 0  // Not currently used
+#if 0 // BU: Not currenly used
```
=> own PR

deadalnix commented on src/requestManager.cpp:
```diff
@@ -429,7 +430,7 @@ void CRequestManager::SendRequests()
       if (now-item.lastRequestTime > blkReqRetryInterval)  // if never requested then lastRequestTime==0 so this will always be true
 	{
-          if (!item.availableFrom.empty())
+         if (!item.availableFrom.empty())
```
formating.

ptschip replied:
done

deadalnix commented on src/rpc/blockchain.cpp:
```diff
@@ -140,7 +140,6 @@ UniValue getblockcount(const UniValue& params, bool fHelp)
             + HelpExampleCli("getblockcount", "")
             + HelpExampleRpc("getblockcount", "")
         );
-
```
formating.

deadalnix commented on src/txdb.cpp:
```diff
@@ -35,21 +33,25 @@ CCoinsViewDB::CCoinsViewDB(size_t nCacheSize, bool fMemory, bool fWipe) : db(Get
 }
 
 bool CCoinsViewDB::GetCoins(const uint256 &txid, CCoins &coins) const {
+    boost::shared_lock<boost::shared_mutex> lock(utxo);
```
Why is that needed ? Isn't the db ACID already ? That'd be worrying if it wasn't.

ptschip replied:
It's a good question...I think i was debugging a problem and put an extra layer of isolation.  I don't think I trust LevelDB entirely so I think I'd like to leave it in place for a little while.  It can probably be taken out but as I said it's a low cost extra layer of isolation which just gives me confidence that when/if I see a problem  then I know it's not  levelDB.

ptschip committed:
More Block Handling updates

-Stop all validation threads by nChainWork

We need to do this when mining our own blocks so we
only kill threads that have less work than our own.

-Add cs_main lock to rpcmining

-Don't evaluate PV.QuitReceived() unless in parallel mode

QuitReceived() gets called for each input that is checked during
connectblock() so we don't want to evaluate it unless we are really
in parallel validation.

-Don't kill your own thread in StopAllValidationThreads by nChainWork

-take out comment in p2p-versionbits-warning

remove old comments and a few format changes.

-Re-enable SENDHEADERS when Xthins is not on

previously had to disable while debugging PV

-Timing fixes for excessive.py

-Re-enable PV.StopAllValidationThreads()

-Keep track of Re-orgs

We need to know if a PV thread has initiated a re-org so we don't
go requesting new blocks while a re-org is in progress.

-Comment code in checkblockindex

ptschip commented:
closing and re-opening PR on the dev branch
